### ISSUE TITLE
Improve garbage collection to detect unreachable actors

### ISSUE THREAD

**Issue by Neverlord** [actor-framework/actor-framework#420]
After the recent discussion on the mailing list, I wonder how CAF could improve its garbage collection. Currently, actors never get destroyed unless they call `quit`. However, semantically it makes no sense to keep unreachable actors alive. Once there is no more reference to an actor and it is not currently running, there is no way for it to ever run again.

A straightforward way to improve garbage collection of actors is as follows:
- Each message in the mailbox of an actor holds 1 reference to the actor
- The scheduler increments the reference count before resuming and decrements it afterwards

This adds 2 operations on the reference count per message and 2 operations per run. Manipulate the reference count for each message would add contention, possibly leading to performance degradation.

An different version could work as follows:
- Reference count of an actor is incremented whenever:
  + a handle to it  is created (new `intrusive_ptr` pointing to the actor)
  + its mailbox changes state from blocked to filled
  + its mailbox fetches new items to the (currently empty) cache
  + `resume` is about to be called by the scheduler
  + it is about to return from `resume` with return code `resume_later`
- Reference count of an actor is decremented whenever:
  + a handle to it is destroyed
  + its mailbox changes state from filled to either blocked or closed
  + the last element from the mailbox cache is taken
  + `resume` is called and the last return code was `resume_later`
  + `resume` returned control back to the scheduler

This version is more complex, but could reduce the frequency of ref count operations for actors with high message throughput. Ideally, we would implement both strategies and benchmark.

**Comment by Neverlord** [comment]
A few observations I had when working on this:

What adds a reference to an actor?
- having a handle of type `actor`, `actor_addr`, or `typed_actor<...>` to it
- monitoring it
- linking to it
- sending it messages (this will cause the scheduler to hold a reference to it)
- publishing it
- composing it via operator*, splice, etc.

I think these points are all obvious, with one exception: monitoring. This means an actor cannot be garbage collected as long as it is monitored. In other words, monitoring an actor even once adds a "permanent" reference count to it. This is only due to the fact that `down_msg` contains an `actor_addr` to the died actor. If we would replace the `actor_addr` with something different that can be compared to an `actor`, `actor_addr`, etc but cannot be "uplifted" using `actor_cast`, we could get rid of this limitation.

Likewise, a link would no longer keep that actor itself alive. An actor would have at least one reference as long as any of its links is still alive, but it wouldn't keep itself alive. Using weak pointers for links would even allow CAF to garbage-collect linked actors that are otherwise unreachable, but this might come at a runtime cost (which we would need to quantify to see if it is measurable / prohibitive).

**Comment by Neverlord** [comment]
Thinking a bit more about this issue, I think weak pointers are the way to go. Otherwise, linking two actors always will create a cycle.

**Comment by Neverlord** [comment]
So, here's another idea that probably solves most issues. `actor_addr` uses weak pointer semantics and `actor`/`typed_actor` use "strong" pointer semantics. An implementation would then "separate" actors into a control block and a data block:

```
             +------------+
             |            | 
             v            |
   +-----------------+--------------+
   |  control block  |  data block  |
   +-----------------+--------------+
   | ref count       | mailbox      |
   | weak ref count  |    .         |
   | actor ID        |    .         |
   | node ID         |    .         |
   +-----------------+------------- +
```

This is much like the design of `std::shared_ptr`, but the control block also contains actor ID and node ID to store the identity of an actor separately (for identifying an actor by its IDs even after it died). We can put both blocks into a single memory region with the data block stored in a `union { }`. This allows us to destroy the data independently from the control block while paying no extra cost for heap indirection (aka pointer chasing). We can even compute the address of the control block via offset from inside an actor to avoid the "weak-pointer-to-self" workaround used by `std::enable_shared_from_this`.

To give `actor_addr` weak pointer semantics would break the current cycles produced by links or when two actors monitor each other. Additionally, the custom memory layout works around (most) additional runtime costs. This also gives `actor_addr` a stronger role, as it is no longer "just" used for type erasure.

**Comment by Neverlord** [comment]
And here is a proof of concept (warning: black magic ahead):

```cpp
namespace {

static constexpr intptr_t cache_line_size = 64;

struct actor_control_block {
  std::atomic<size_t> strong_refs;
  std::atomic<size_t> weak_refs;
  actor_id aid;
  node_id nid;
};

template <class T>
struct embedded_actor {
  actor_control_block ctrl;
  char pad[cache_line_size - sizeof(actor_control_block)];
  union { T data; };
  embedded_actor() {
    new (&data) T();
  }
  ~embedded_actor() {
    data.~T();
  }
  embedded_actor(const embedded_actor&) = delete;
  embedded_actor& operator=(const embedded_actor&) = delete;
};

class actor_dummy_base {
public:
  virtual ~actor_dummy_base() {
    // nop
  }

  actor_control_block* ctrl() {
    return reinterpret_cast<actor_control_block*>(reinterpret_cast<intptr_t>(this) - cache_line_size);
  }

  // allow placement new
  inline void* operator new(std::size_t, void* ptr) {
    return ptr;
  }

private:
  // prohibit custom heap allocation
  void* operator new(size_t);
  void* operator new[](size_t);
};

template <class T>
embedded_actor<T>* new_embedded_actor() {
  using type = embedded_actor<T>;
  //static_assert(std::is_standard_layout<type>::value,
  //              "embedded_actor<T> violates standard layout");
  static_assert(offsetof(type, data) == cache_line_size,
                "data is not at cache line size boundary");
  return new type;
}

class actor_dummy : public actor_dummy_base {
public:
  ~actor_dummy() {
    // nop
  }
};

} // namespace <anonymous>

CAF_TEST(actor_ctrl) {
  auto tst = new_embedded_actor<actor_dummy>();
  tst->ctrl.aid = 42;
  CAF_CHECK(tst->ctrl.aid == 42);
  CAF_CHECK(tst->data.ctrl() == &(tst->ctrl));
  delete tst;
}
```

Of course this means users must never allocate actors themselves, which is done by new operators other than placement new private. Technically, users *could* still put an actor onto the stack, but that would be a rather silly thing to do in the first place.

**Comment by Neverlord** [comment]
And here's some graffle for the current design:

![refcounting](https://cloud.githubusercontent.com/assets/651128/14380483/f81163a6-fd7f-11e5-86ff-353c99fe4740.png)

The `strong_actor_ptr` does contain only a single pointer to the actor control block (which notably isn't a template). When obtaining an `abstract_actor*` pointer from the `strong_actor_pointer`, it simply calculates the location of the actor via pointer arithmetic. This requires that `static_cast` and `reinterpret_cast` perform the same operation when casting from `T*` to `abstract_actor*` or vice versa, which is checked via `static_assert`. In practice, the only limitation users have in this design is that actors must not use virtual inheritance.

Notably, the new design didn't change the performance of the `actor_creation` and `mixed_case` benchmarks on 64 cores. So I think this is ready make its way into the actor system topic branch (and `develop` eventually).

**Comment by Neverlord** [comment]
Garbage collecting actors locally is implemented via 47dd7ca. However, tracking references in a distributed setting is not properly solved yet. The current "solution" is to have the actor registry store `strong_actor_ptr`. This avoids deleting an actor that could still be referenced on another node, but keeps actors alive indefinitely when they're not. This was the case previously anyways, but probably we want to re-visit this issue at some point in the future to have "proper" garbage collection in distributed actor systems.

**Comment by Neverlord** [comment]
And here a few follow-up tasks for this issue. Help is much appreciated.

[] re-evaluate the implementation of actor decorators w.r.t. proper garbage collection
[] kick out all code in unit tests that kills testee actors before leaving the `CAF_TEST` function
[] properly document new behavior and types (I'm already working on that)
